test/unit/factorization/mc/test_compute_t_factor.cpp

```diff
@@ -77,6 +80,100 @@ void is_orthogonal(const MatrixLocal<const T>& matrix) {
   CHECK_MATRIX_NEAR(eye, ortho, 0, error);
 }
 
+template <class T>
+std::tuple<hpx::shared_future<dlaf::common::internal::vector<T>>, MatrixLocal<T>> computeTauH(
+    const dlaf::SizeType k, const dlaf::TileElementSize block_size, const MatrixLocal<T>& v) {
+  const dlaf::SizeType m = v.size().rows();
+
+  // compute taus and H_exp
+  dlaf::common::internal::vector<T> taus;
+  taus.reserve(k);
+
+  MatrixLocal<T> h_expected({m, m}, block_size);
+  set(h_expected, preset_eye<T>);
+
+  for (auto j = 0; j < k; ++j) {
+    const dlaf::SizeType reflector_size = m - j;
+
+    const T* data_ptr = v.ptr({j, j});
+    const auto norm = blas::nrm2(reflector_size, data_ptr, 1);
+    const T tau = 2 / std::pow(norm, static_cast<decltype(norm)>(2));
+
+    taus.push_back(tau);
+    MatrixLocal<T> h_i({reflector_size, reflector_size}, block_size);
+    set(h_i, preset_eye<T>);
+
+    // Hi = (I - tau . v . v*)
+    // clang-format off
+    blas::ger(blas::Layout::ColMajor,
+        reflector_size, reflector_size,
+        -tau,
+        v.ptr({j, j}), 1,
+        v.ptr({j, j}), 1,
+        h_i.ptr(), h_i.ld());
+    // clang-format on
+
+    // H_exp[:, j:] = H_exp[:, j:] . Hi
+    const dlaf::GlobalElementIndex h_offset{0, j};
+    MatrixLocal<T> workspace({h_expected.size().rows(), h_i.size().cols()}, h_i.blockSize());
+
+    // clang-format off
+    blas::gemm(blas::Layout::ColMajor,
+	blas::Op::NoTrans, blas::Op::NoTrans,
+	h_expected.size().rows(), h_i.size().cols(), h_i.size().rows(),
+	1,
+	h_expected.ptr(h_offset), h_expected.ld(),
+	h_i.ptr(), h_i.ld(),
+	0,
+	workspace.ptr(), workspace.ld());
+    // clang-format on
+    std::copy(workspace.ptr(), workspace.ptr() + workspace.size().linear_size(),
+              h_expected.ptr(h_offset));
+  }
+  hpx::shared_future<dlaf::common::internal::vector<T>> taus_input = hpx::make_ready_future(taus);
+
+  return std::make_tuple(taus_input, std::move(h_expected));
+}
+
+template <class T>
+MatrixLocal<T> computeHres(const dlaf::SizeType k, const dlaf::TileElementSize block_size,
+                           const dlaf::GlobalElementSize tile_size, Matrix<T, Device::CPU>& t_output,
+                           const dlaf::LocalTileIndex t_idx, const MatrixLocal<T>& v) {
```

Comment:
- `block_size = v.blockSize()` no need to pass it as extra parameter.

- `tile_size`? You are passing the size of the orthogonal matrix.
Anyway, this is simply {m, m}, therefore there is no need to pass it as an argument.

- Simply pass the tile t (`const Tile<const T, Device::CPU>& t`). Where it is stored is not relevant for this function.

- Again, elements of v should be const.